### PR TITLE
Adam selectedrows and scatter functors

### PR DESCRIPTION
**Pull request by typhoonzero**
Fix https://github.com/PaddlePaddle/Paddle/issues/7019

Start: 2017.12.27
Due: 2018.1.3

### CODE REVIEW

File: paddle/operators/adam_op.h [hunk 1]
```diff
@@ -79,6 +79,70 @@ struct AdamFunctor {
...
+        row_numel_(row_numel) {}
+
+  inline HOSTDEVICE void operator()(size_t i) const {
+    for (int64_t j = 0; j < row_numel_; ++j) {
+      T g = grad_[i * row_numel_ + j];
+      T mom1 = moment1_[rows_[i] * row_numel_ + j];
+      T mom2 = moment2_[rows_[i] * row_numel_ + j];
+      T lr = *lr_;
+      T beta1_pow = *beta1_pow_;
+      T beta2_pow = *beta2_pow_;
```
Comment (chengduoZH) [inline paddle/operators/adam_op.h]:
```
T lr = *lr_;
T beta1_pow = *beta1_pow_;
T beta2_pow = *beta2_pow_;
```
For GPU, these operations are always call the data from memory to register?

File: paddle/operators/adam_op.h [hunk 2]
```diff
@@ -79,6 +79,70 @@ struct AdamFunctor {
...
+      T p = param_[rows_[i] * row_numel_ + j];
+
+      lr *= sqrt(1 - beta2_pow) / (1 - beta1_pow);
+      mom1 = beta1_ * mom1 + (1 - beta1_) * g;
+      mom2 = beta2_ * mom2 + (1 - beta2_) * g * g;
+      p -= lr * (mom1 / (sqrt(mom2) + epsilon_));
+      // IMPORTANT:
+      // FIXME(typhoonzero): row id may be duplicate
```
Comment (QiJune) [inline paddle/operators/adam_op.h]: Sparse Adagrad merges row id first. Please refer to https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/adagrad_op.cc#L107.

Comment (typhoonzero) [reply paddle/operators/adam_op.h]: Thanks! Will update this.

File: paddle/operators/math/selected_rows_functor.cu [hunk 3]
```diff
@@ -222,6 +224,157 @@ template struct SelectedRowsAddToTensor<platform::CUDADeviceContext, float>;
...
+                               T* out, const int64_t* out_rows,
+                               size_t out_rows_size, int64_t row_numel) {
+  const int ty = blockIdx.y;
+  int tid = threadIdx.x;
+  __shared__ size_t out_idx;
+
+  if (tid == 0) {
+    for (size_t i = 0; i < out_rows_size; i++) {
+      if (input_rows[ty] == out_rows[i]) {
+        out_idx = i;
```
Comment (QiJune) [inline paddle/operators/math/selected_rows_functor.cu]: We can add a break here.

Comment (typhoonzero) [reply paddle/operators/math/selected_rows_functor.cu]: Sure. I'll try to fix in next PR since I'm considering a way to simplify the functors.

**Pull request merged**